// xmlbuilder/lib/XMLBuilder.js
// Generated by CoffeeScript 1.10.0
(function() {
  var XMLBuilder, XMLElement, XMLStringWriter, XMLStringifier, isPlainObject;

  isPlainObject = require('lodash/isPlainObject');

  XMLStringifier = require('./XMLStringifier');

  XMLElement = require('./XMLElement');

  XMLStringWriter = require('./XMLStringWriter');

  module.exports = XMLBuilder = (function() {
    function XMLBuilder(name, options) {
      var root, temp;
      if (name == null) {
        throw new Error("Root element needs a name");
      }
      if (options == null) {
        options = {};
      }
      if (!options.writer) {
        options.writer = new XMLStringWriter();
      }
      this.options = options;
      this.stringify = new XMLStringifier(options);
      temp = new XMLElement(this, 'doc');
      root = temp.element(name);
      root.isRoot = true;
      root.documentObject = this;
      this.rootObject = root;
      if (!options.headless) {
        root.declaration(options);
        if ((options.pubID != null) || (options.sysID != null)) {
          root.doctype(options);
        }
      }
    }

    XMLBuilder.prototype.dec = function() {
      return this.xmldec;
    };

    XMLBuilder.prototype.dtd = function() {
      return this.doctype;
    };

    XMLBuilder.prototype.root = function() {
      return this.rootObject;
    };

    XMLBuilder.prototype.end = function(writer) {
      var writerOptions;
      if (!writer) {
        writer = this.options.writer;
      } else if (isPlainObject(writer)) {
        writerOptions = writer;
        writer = this.options.writer.set(writerOptions);
      }
      return writer.document(this);
    };

    XMLBuilder.prototype.toString = function(options) {
      return this.options.writer.set(options).document(this);
    };

    return XMLBuilder;

  })();

}).call(this);
